### PR TITLE
keyboard shortcut rewrite + shortcut context + maskEditor shortcuts

### PR DESCRIPTION
**Pull request by trsommer**
- Added context system for keyboard shortcuts to limit their scope to specific UI areas
- Implemented keyboard controls for the mask editor
- Created UI for easy shortcut customization
- added setting to change stepSize of new brush increase/brush decrease shortcut
- added undo & redo to coreKeybindings

image

Future Work needed:
- remove bypassed typescript strict errors in 2 files
- restrict reset keyboard shortcut button to current context
- fix playwright tests

### CODE REVIEW

File: src/constants/coreSettings.ts [hunk 1]
```diff
@@ -410,18 +410,11 @@ export const CORE_SETTINGS: SettingParams[] = [
     versionAdded: '1.3.5'
   },
   {
-    id: 'Comfy.Keybinding.UnsetBindings',
```
Comment (huchenlei) [inline src/constants/coreSettings.ts:413]: Why do we need to nuke old settings. Doing so would break all existing users' keybinding setup.

**Review by huchenlei — CHANGES_REQUESTED** [review]
Overall I don't think the design of having the context is necessary here. The selector field in the current definition should be reworked to manage the invocation instead.

This PR is doing too many things together which makes it very hard to review. Please consider break it down to separate PRs:
- keyboard shortcut rewrite
- shortcut context
- maskEditor shortcuts

File: src/constants/coreSettings.ts (hunk 1 above)
Comment (trsommer) [reply src/constants/coreSettings.ts:413]: Good point. When I have more time, I can build a migration function

**Comment by trsommer** [comment]
> Overall I don't think the design of having the context is necessary here. The selector field in the current definition should be reworked to manage the invocation instead.
> 
> This PR is doing too many things together which makes it very hard to review. Please consider break it down to separate PRs:
> 
>     * keyboard shortcut rewrite
> 
>     * shortcut context
> 
>     * maskEditor shortcuts

I agree that this pull request is big and probably should be split up. But the addition of the context required deep changes to the whole shortcut system. That's why I decided to completely rewrite it as well to build a system that is built for the contexts instead of adapting the old system that was not designed to support contexts. The Target Selector in its current form at least is not sufficient because it is bound to a specific CSS selector and doesn't support the same combo in 2 different scenarios. (e.g. undo). The rewrite was necessary because the old system used the serialized keybinding as the index for the keybinding. That system obviously doesn't support multiple same combos.
I agree that nuking user set keybindings is not necessary. 
I would propose the following:
1. I will build a migration function so no user keybindings are lost
2. I will also split this pull request up so that the mask editor shortcuts are in a separate PR
3. I will add jsdoc to help with the review process.

**Comment by huchenlei** [comment]
@trsommer I invited you as collaborator of this repo. This will allow you create branches directly in the repo for PR so some github actions such as locale update can be correctly triggered.

**Comment by charhubbie** [comment]
Where is this in the new version? I absolutely hate there's no way to save or leave the interface now with the new UI. although the alt click is good to increase brush size. Before i could hit ''esc'' to leave the screen, or ''enter'' to save my changes.

**Pull request closed without merging**